File: python/flyql/generators/clickhouse/column.py
"""ClickHouse-dialect Column. Construct via ``Column(name=, _type=,
...)``. Public surface is ``name`` / ``raw_identifier`` / ``values`` /
``display_name`` plus the ``raw_type`` and ``flyql_type`` properties.
The ``flyql_type`` is computed at construction from the raw DB type via
:func:`normalize_clickhouse_type`."""

import re
from typing import Dict, List, Optional

from flyql.flyql_type import Type

REGEX = {
    "wrapper": re.compile(
        r"^(nullable|lowcardinality|simpleaggregatefunction|aggregatefunction)\s*\(\s*(.+)\)"
    ),
    Type.String: re.compile(r"^(varchar|char|fixedstring)\s*\(\s*\d+\s*\)"),
    Type.Int: re.compile(
        r"^(tinyint|smallint|mediumint|int|integer|bigint)\s*\(\s*\d+\s*\)"
    ),
    Type.Float: re.compile(r"^(decimal|numeric|dec)\d*\s*\(\s*\d+\s*(,\s*\d+)?\s*\)"),
    Type.Date: re.compile(r"^datetime64\s*\(\s*\d+\s*(,\s*.+)?\s*\)"),
    Type.Array: re.compile(r"^array\s*\("),
    Type.Map: re.compile(r"^map\s*\("),
    Type.Struct: re.compile(r"^tuple\s*\("),
    Type.JSON: re.compile(r"^json\s*\("),
}

# Lookup table for raw ClickHouse DB type names that don't require regex
# matching. CH renames: tuple→Struct, interval→Duration, geometry/special
# collapse into Unknown.
FLYQL_TYPE_TO_CLICKHOUSE_TYPES: Dict[Type, set] = {
    Type.String: {
        "string",
        "fixedstring",
        "longtext",
        "mediumtext",
        "tinytext",
        "text",
        "longblob",
        "mediumblob",
        "tinyblob",
        "blob",
        "varchar",
        "char",
        "char large object",
        "char varying",
        "character",
        "character large object",
        "character varying",
        "nchar large object",
        "nchar varying",
        "national character large object",
        "national character varying",
        "national char varying",
        "national character",
        "national char",
        "binary large object",
        "binary varying",
        "clob",
        "nchar",
        "nvarchar",
        "varchar2",
        "binary",
        "varbinary",
        "bytea",
        "uuid",
        "ipv4",
        "ipv6",
        "enum8",
        "enum16",
    },
    Type.Int: {
        "int8",
        "int16",
        "int32",
        "int64",
        "int128",
        "int256",
        "uint8",
        "uint16",
        "uint32",
        "uint64",
        "uint128",
        "uint256",
        "tinyint",
        "smallint",
        "mediumint",
        "int",
        "integer",
        "bigint",
        "tinyint signed",
        "tinyint unsigned",
        "smallint signed",
        "smallint unsigned",
        "mediumint signed",
        "mediumint unsigned",
        "int signed",
        "int unsigned",
        "integer signed",
        "integer unsigned",
        "bigint signed",
        "bigint unsigned",
        "int1",
        "int1 signed",
        "int1 unsigned",
        "byte",
        "signed",
        "unsigned",
        "bit",
        "set",
        "time",
    },
    Type.Float: {
        "float32",
        "float64",
        "float",
        "double",
        "double precision",
        "real",
        "decimal",
        "decimal32",
        "decimal64",
        "decimal128",
        "decimal256",
        "dec",
        "numeric",
        "fixed",
        "single",
    },
    Type.Bool: {"bool", "boolean"},
    Type.Date: {
        "date",
        "date32",
        "datetime",
        "datetime32",
        "datetime64",
        "timestamp",
        "year",
    },
    Type.Duration: {
        "intervalday",
        "intervalhour",
        "intervalmicrosecond",
        "intervalmillisecond",
        "intervalminute",
        "intervalmonth",
        "intervalnanosecond",
        "intervalquarter",
        "intervalsecond",
        "intervalweek",
        "intervalyear",
    },
    Type.Unknown: {
        "geometry",
        "point",
        "polygon",
        "multipolygon",
        "linestring",
        "ring",
        "nothing",
        "nested",
        "object",
        "dynamic",
        "variant",
    },
    Type.JSON: {"json"},
}


def normalize_clickhouse_type(ch_type: str) -> Type:
    """Map a raw ClickHouse DB type string to its canonical
    :class:`flyql.flyql_type.Type`. Unknown raw types map to ``Type.Unknown``."""
    if not ch_type or not isinstance(ch_type, str):
        return Type.Unknown

    normalized = ch_type.strip().lower()

    if normalized == "jsonstring":
        return Type.JSONString

    match = REGEX["wrapper"].match(normalized)
    if match:
        normalized = match.group(2).strip()

    if REGEX[Type.String].match(normalized):
        return Type.String
    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.String]:
        return Type.String

    if REGEX[Type.Int].match(normalized):
        return Type.Int
    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Int]:
        return Type.Int

    if REGEX[Type.Float].match(normalized):
        return Type.Float
    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Float]:
        return Type.Float

    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Bool]:
        return Type.Bool

    if REGEX[Type.Date].match(normalized):
        return Type.Date
    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Date]:
        return Type.Date

    if REGEX[Type.JSON].match(normalized):
        return Type.JSON
    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.JSON]:
        return Type.JSON

    if REGEX[Type.Array].match(normalized):
        return Type.Array

    if REGEX[Type.Map].match(normalized):
        return Type.Map

    if REGEX[Type.Struct].match(normalized):
        return Type.Struct

    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Unknown]:
        return Type.Unknown

    if normalized in FLYQL_TYPE_TO_CLICKHOUSE_TYPES[Type.Duration]:
        return Type.Duration

    return Type.Unknown


def _escape_identifier(name: str) -> str:
    needs_quoting = name[:1].isdigit() or any(
        not (c.isalnum() or c == "_") for c in name
    )
    if not needs_quoting:
        return name
    escaped = name.replace("`", "``")
    return f"`{escaped}`"


class Column:
    """Opaque ClickHouse-dialect column. ``flyql_type`` is the primary
    dispatch input for generator code; ``raw_type`` is the fallback for
    DDL-level inspection."""

    def __init__(
        self,
        name: str,
        _type: str,
        values: Optional[List[str]] = None,
        display_name: str = "",
        raw_identifier: str = "",
    ):
        # Public attributes
        self.name = _escape_identifier(name)
        self.match_name = name
        self.values: List[str] = values or []
        self.display_name = display_name
        self.raw_identifier = raw_identifier
        # Internal type fields
        self._raw_type = _type
        self._flyql_type: Type = normalize_clickhouse_type(_type)

    @property
    def raw_type(self) -> str:
        return self._raw_type

    @property
    def flyql_type(self) -> Type:
        return self._flyql_type

    def with_raw_identifier(self, identifier: str) -> "Column":
        self.raw_identifier = identifier
        return self

File: javascript/packages/flyql/src/generators/clickhouse/column.js
import { Type } from '../../flyql_type.js'
import { FlyqlError } from '../../core/exceptions.js'

const wrapperRegex = /^(nullable|lowcardinality|simpleaggregatefunction|aggregatefunction)\(/i

const typeRegexes = {
    [Type.String]: /^(varchar|char|fixedstring)\s*\(\s*\d+\s*\)/i,
    [Type.Int]: /^(tinyint|smallint|mediumint|int|integer|bigint)\s*\(\s*\d+\s*\)/i,
    [Type.Float]: /^(decimal|numeric|dec)\d*\s*\(\s*\d+\s*(,\s*\d+)?\s*\)/i,
    [Type.Date]: /^datetime64\s*\(\s*\d+\s*(,\s*.+)?\s*\)/i,
    [Type.Array]: /^array\s*\(/i,
    [Type.Map]: /^map\s*\(/i,
    [Type.Struct]: /^tuple\s*\(/i,
    [Type.JSON]: /^json\s*\(/i,
}

const flyqlTypeToClickHouseTypes = {
    [Type.String]: new Set([
        'string',
        'fixedstring',
        'longtext',
        'mediumtext',
        'tinytext',
        'text',
        'longblob',
        'mediumblob',
        'tinyblob',
        'blob',
        'varchar',
        'char',
        'char large object',
        'char varying',
        'character',
        'character large object',
        'character varying',
        'nchar large object',
        'nchar varying',
        'national character large object',
        'national character varying',
        'national char varying',
        'national character',
        'national char',
        'binary large object',
        'binary varying',
        'clob',
        'nchar',
        'nvarchar',
        'varchar2',
        'binary',
        'varbinary',
        'bytea',
        'uuid',
        'ipv4',
        'ipv6',
        'enum8',
        'enum16',
    ]),
    [Type.Int]: new Set([
        'int8',
        'int16',
        'int32',
        'int64',
        'int128',
        'int256',
        'uint8',
        'uint16',
        'uint32',
        'uint64',
        'uint128',
        'uint256',
        'tinyint',
        'smallint',
        'mediumint',
        'int',
        'integer',
        'bigint',
        'tinyint signed',
        'tinyint unsigned',
        'smallint signed',
        'smallint unsigned',
        'mediumint signed',
        'mediumint unsigned',
        'int signed',
        'int unsigned',
        'integer signed',
        'integer unsigned',
        'bigint signed',
        'bigint unsigned',
        'int1',
        'int1 signed',
        'int1 unsigned',
        'byte',
        'signed',
        'unsigned',
        'bit',
        'set',
        'time',
    ]),
    [Type.Float]: new Set([
        'float32',
        'float64',
        'float',
        'double',
        'double precision',
        'real',
        'decimal',
        'decimal32',
        'decimal64',
        'decimal128',
        'decimal256',
        'dec',
        'numeric',
        'fixed',
        'single',
    ]),
    [Type.Bool]: new Set(['bool', 'boolean']),
    [Type.Date]: new Set(['date', 'date32', 'datetime', 'datetime32', 'datetime64', 'timestamp', 'year']),
    [Type.Duration]: new Set([
        'intervalday',
        'intervalhour',
        'intervalmicrosecond',
        'intervalmillisecond',
        'intervalminute',
        'intervalmonth',
        'intervalnanosecond',
        'intervalquarter',
        'intervalsecond',
        'intervalweek',
        'intervalyear',
    ]),
    [Type.Unknown]: new Set([
        'geometry',
        'point',
        'polygon',
        'multipolygon',
        'linestring',
        'ring',
        'nothing',
        'nested',
        'object',
        'dynamic',
        'variant',
    ]),
    [Type.JSON]: new Set(['json']),
}

export function normalizeClickHouseType(chType) {
    if (!chType) return Type.Unknown

    let normalized = chType.trim().toLowerCase()

    if (normalized === 'jsonstring') return Type.JSONString

    const wrapperMatch = normalized.match(wrapperRegex)
    if (wrapperMatch) {
        const afterKeyword = normalized.slice(wrapperMatch[0].length)
        const lastParen = afterKeyword.lastIndexOf(')')
        normalized = (lastParen >= 0 ? afterKeyword.slice(0, lastParen) : afterKeyword).trim()
    }

    if (typeRegexes[Type.String].test(normalized)) return Type.String
    if (flyqlTypeToClickHouseTypes[Type.String].has(normalized)) return Type.String

    if (typeRegexes[Type.Int].test(normalized)) return Type.Int
    if (flyqlTypeToClickHouseTypes[Type.Int].has(normalized)) return Type.Int

    if (typeRegexes[Type.Float].test(normalized)) return Type.Float
    if (flyqlTypeToClickHouseTypes[Type.Float].has(normalized)) return Type.Float

    if (flyqlTypeToClickHouseTypes[Type.Bool].has(normalized)) return Type.Bool

    if (typeRegexes[Type.Date].test(normalized)) return Type.Date
    if (flyqlTypeToClickHouseTypes[Type.Date].has(normalized)) return Type.Date

    if (typeRegexes[Type.JSON].test(normalized)) return Type.JSON
    if (flyqlTypeToClickHouseTypes[Type.JSON].has(normalized)) return Type.JSON

    if (typeRegexes[Type.Array].test(normalized)) return Type.Array

    if (typeRegexes[Type.Map].test(normalized)) return Type.Map

    if (typeRegexes[Type.Struct].test(normalized)) return Type.Struct

    if (flyqlTypeToClickHouseTypes[Type.Unknown].has(normalized)) return Type.Unknown

    if (flyqlTypeToClickHouseTypes[Type.Duration].has(normalized)) return Type.Duration

    return Type.Unknown
}

function escapeIdentifier(name) {
    if (/^[0-9]/.test(name) || /[^a-zA-Z0-9_]/.test(name)) {
        const escaped = name.replace(/`/g, '``')
        return `\`${escaped}\``
    }
    return name
}

/**
 * Opaque ClickHouse-dialect Column. Construct via `newColumn({...})` or `new Column({...})`.
 * The flyql semantic type is computed at construction via `normalizeClickHouseType`.
 */
export class Column {
    constructor(opts) {
        if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
            throw new FlyqlError(`Column: expected an options object ({ name, type, values? }); got ${typeof opts}`)
        }
        const { name, type, values = null, displayName = '', rawIdentifier = '' } = opts
        if (typeof name !== 'string' || !name) {
            throw new FlyqlError(`Column: 'name' must be a non-empty string`)
        }
        if (typeof type !== 'string') {
            throw new FlyqlError(
                `Column: 'type' must be a raw-type string (e.g. 'String', 'jsonstring'); got ${typeof type}`,
            )
        }

        this.name = escapeIdentifier(name)
        this.values = values || []
        this.displayName = displayName
        this.rawIdentifier = rawIdentifier
        this.matchName = name
        this._rawType = type
        this._flyqlType = normalizeClickHouseType(type)
    }

    rawType() {
        return this._rawType
    }

    flyqlType() {
        return this._flyqlType
    }

    /** @deprecated Pass `rawIdentifier` in the options object to `newColumn()` / `new Column()` instead. Slated for removal in a follow-up. */
    withRawIdentifier(identifier) {
        this.rawIdentifier = identifier
        return this
    }
}

export function newColumn(opts) {
    if (opts == null || typeof opts !== 'object' || Array.isArray(opts)) {
        throw new FlyqlError(`newColumn: expected an options object ({ name, type, values? }); got ${typeof opts}`)
    }
    return new Column(opts)
}
